### PR TITLE
Core/Spells: Fixed warlock's Banish cancel if target was already banished

### PR DESCRIPTION
**Pull request by ForesterDev**
**Changes proposed:**

-  Core/Spells: Add SpellMissInfo argument to BeforeHit hooks and call them also when the spell doesn't hit. (#17613)

(cherry picked from commit 8ff5b35)

-  Core/Spells: Fixed warlock's Banish cancel if target was already banished (#17614)

(cherry picked from commit 4587b5d)

**Target branch(es):**

- [x] 3.3.5

**Tests performed:** (Does it build, tested in-game, etc.) Build, tested

### CODE REVIEW

File: src/server/game/Spells/SpellInfo.cpp [hunk 1]
```diff
@@ -1346,7 +1346,7 @@ bool SpellInfo::CanPierceImmuneAura(SpellInfo const* auraSpellInfo) const
         // ...but not these (Divine shield, Ice block, Cyclone and Banish for example)
         if (auraSpellInfo->Mechanic != MECHANIC_IMMUNE_SHIELD &&
                auraSpellInfo->Mechanic != MECHANIC_INVULNERABILITY &&
-               (auraSpellInfo->Mechanic != MECHANIC_BANISH || (IsRankOf(auraSpellInfo) && auraSpellInfo->Dispel != DISPEL_NONE))) // Banish shouldn't be immune to itself, but Cyclone should
+               auraSpellInfo->Mechanic != MECHANIC_BANISH)
```
Comment (ForesterDev) [inline src/server/game/Spells/SpellInfo.cpp]: Wihout this change, banish pierce throw banish, so script won't work

Comment (jackpoz) [reply src/server/game/Spells/SpellInfo.cpp]: is this needed in master too ?

Comment (ForesterDev) [reply src/server/game/Spells/SpellInfo.cpp]: Don't test in master, but I think - yes

Comment (jackpoz) [reply src/server/game/Spells/SpellInfo.cpp]: could you check that there is no regression about https://github.com/TrinityCore/TrinityCore/issues/18479 and also that Cyclone still works the way it's supposed to ?

Comment (ForesterDev) [reply src/server/game/Spells/SpellInfo.cpp]: image
Works fine

**Comment by ccrs** [comment]
I know this is just a port from master, but this implementation is so weird..

also `Spell::_handle_immediate_phase()` is called before `Spell::DoProcessTargetContainer` inside `Spell::handle_immediate()`..
that means, this is called before any script hook
```c++
    // handle effects with SPELL_EFFECT_HANDLE_HIT mode
    for (uint32 j = 0; j < MAX_SPELL_EFFECTS; ++j)
    {
        // don't do anything for empty effect
        if (!m_spellInfo->Effects[j].IsEffect())
            continue;

        // call effect handlers to handle destination hit
        HandleEffects(nullptr, nullptr, nullptr, j, SPELL_EFFECT_HANDLE_HIT);
    }
```

so... for `SPELL_EFFECT_HANDLE_HIT` the default handler cannot be overriden with `BeforeHit` script hook... ? makes any sense?
@Shauren

**Pull request merged**